### ISSUE TITLE
Curious Performance Bug with Moduli and product

### ISSUE THREAD

**Issue by CarlEdman** [cartazio/arithmoi#160]
Compare these two codes:
`main = print $ product [1..10^7::Mod 1000000007]`
and:
`main = print $ foldl' (*) (1::Mod 1000000007) [1..10^7]`
These should be equivalent and indeed they both produce the same (correct) result.  

But the runtimes are dramatically different.   I've attached the full results below, but the highlights are:

- The foldl' version allocates 1.4 GByte on the heap, the product version 2.3 GByte

- The foldl' version copies only 100 kByte during GC, the product version almost everything it allocated.

- As a consequence, the foldl' version has about 0 GC time, the product version 2s of GC time.

- In addition, the product version takes about twice as much MUT time as the foldl' version.

- Net result: The product version takes almost 8 times longer.

This may or may not be related to arithmoi.  If the conclusion is that it isn't please feel free to close this issue.

I was actually quite impressed that the foldl' Mod version had almost exactly the same performance as the explicit version using Integers.  In the past I've tried other Type-Magic mod packages, but the performance was always so much worse that I stuck with the express code wherever performance matters.  No more!

But note that the express version with a modulus smaller than half the word size of the machine (which is still quite a lot on a 64-bit machine), so you don't need to do anything complicated on every multiplication is **still three times faster** than the Mod version.  If some type magic wizard could add some code to use Words rather than Integers for sufficiently small moduli, it would still be a big win.

foldl' version:
```1,440,050,352 bytes allocated in the heap
         118,912 bytes copied during GC
          42,928 bytes maximum residency (2 sample(s))
          26,704 bytes maximum slop
               2 MB total memory in use (0 MB lost due to fragmentation)

                                     Tot time (elapsed)  Avg pause  Max pause
  Gen  0      1381 colls,     0 par    0.000s   0.003s     0.0000s    0.0001s
  Gen  1         2 colls,     0 par    0.000s   0.001s     0.0004s    0.0006s

  INIT    time    0.000s  (  0.001s elapsed)
  MUT     time    0.328s  (  0.348s elapsed)
  GC      time    0.000s  (  0.004s elapsed)
  EXIT    time    0.000s  (  0.000s elapsed)
  Total   time    0.328s  (  0.352s elapsed)

  %GC     time       0.0%  (1.0% elapsed)

  Alloc rate    4,388,724,882 bytes per MUT second

  Productivity 100.0% of total user, 98.8% of total elapsed
```

product version:
```
   2,341,584,064 bytes allocated in the heap
   2,284,138,544 bytes copied during GC
     679,162,552 bytes maximum residency (11 sample(s))
       6,051,896 bytes maximum slop
            1136 MB total memory in use (0 MB lost due to fragmentation)

                                     Tot time (elapsed)  Avg pause  Max pause
  Gen  0      1926 colls,     0 par    1.062s   1.022s     0.0005s    0.0010s
  Gen  1        11 colls,     0 par    0.859s   1.657s     0.1506s    0.6272s

  INIT    time    0.000s  (  0.000s elapsed)
  MUT     time    0.562s  (  0.608s elapsed)
  GC      time    1.922s  (  2.679s elapsed)
  EXIT    time    0.000s  (  0.000s elapsed)
  Total   time    2.484s  (  3.287s elapsed)

  %GC     time      77.4%  (81.5% elapsed)

  Alloc rate    4,162,816,113 bytes per MUT second

  Productivity  22.6% of total user, 18.5% of total elapsed
```

**Comment by CarlEdman** [comment]
FWIW, I forgot to mention that my environment is Haskell Platform 8.4.3 (i.e., ghc 8.4.3), I'm running arithmoi 0.8.0.0, compile with flags `-O2 -optc-O3 -funfolding-use-threshold=16 -fexcess-precision -Wall -Wno-type-defaults -Wno-unused-imports -Wno-unused-top-binds -rtsopts`, and run with flags `+RTS -M12G -s`.

**Comment by Bodigrim** [comment]
`product` is defined in `base` as `foldl (*)`, so there is not much control over its behaviour from library's side. Sometimes GHC can infer that the data is strict, making `foldl` effectively equivalent to `foldl'`, but for some reason it does not for `Mod`. Putting bangs might help.

Could you please check your observations (expecially comparison against `Word`) in GHC 8.6? Internally `Mod` is represented by `Natural` and `Natural` got much better internal support in GHC 8.6.

One can define `SmallMod (m :: Nat) = SmallMod Word` instead of `Mod (m :: Nat) = Mod Natural`, but there is an overhead still, because on type level (`m`) only arbitrarily-long integer literals are supported.

**Comment by CarlEdman** [comment]
By the way, please accept my apologies if I sound over-critical.  Actually matching the performance of Integer with Mod is very impressive and good enough that I'll switch to using it for all but the most performance-sensitive of my code.  That the Word code is still three times faster is testament to the quality of the GHC optimizer; that code apparently does no heap allocation, however ephemeral, at all and appears to be compiled to something close to the equivalent assembly instructions.

**Comment by Bodigrim** [comment]
No need to apologise, everything is fine. Sorry, I have a limited bandwidth these days.

**Comment by Bodigrim** [comment]
I managed to help GHC recognise that multiplication is strict in both arguments, so now `product` and `foldl'` take the same time.

**Comment by Bodigrim** [comment]
I made an attempt to grab some low-hanging fruits. @CarlEdman could you please benchmark?

**Comment by CarlEdman** [comment]
Thanks much! 

I tried installing the github head version of arithmoi on my Haskell machine, but between it being a Windows machine running Haskell Platform and me being about as far from a cabal wizard as it is possible to be, I failed with some apparent incompatibilities in the semirings dependency.  

As it looks like I'm too fumble-fingered to install Haskell package except from hackage release, I will have to defer benchmarking until there is one.  In the meantime, let me close the issue.

**Comment by Bodigrim** [comment]
I benchmarked it roughly on my machine. It used to be 4.1 sec for `foldl'` of `Mod 1000000007` and 1.1 sec for `Word`. Now it takes 2.6 sec for `Mod 1000000007` (both for `foldl'` and for `product`). 

I also executed a test with `newtype SmallMod (m :: Nat) = Mod Word` and it took 1.7 sec. It seems that 1.5x slowdown is caused purely by keeping modulo on type level in a boxed form (unfortunately, there are no type-level `Word`, only `Natural` in GHC). 

I am not entirely convinced that possible improvement from 2.6 to 1.7 src justifies forking the implementation.
